### PR TITLE
Install command usability improvements

### PR DESCRIPTION
**Pull request by mwleeds**
These are a few commits aimed at improving the usability of the Flatpak CLI and addressing some of the issues raised in https://github.com/flatpak/flatpak/issues/1258. These commits do three things:

- Allow the `REF`(s) specified in the `install` command to be "similar" to the actual ref (two characters different or a substring) rather than an exact match, asking the user for confirmation when it's not an exact match
- Allow the `REMOTE` positional argument for the `install` command to be omitted, in which case Flatpak tries to determine which remote to use, and again asks the user for confirmation
- Allow "yes/no" prompts to default to yes when the user just presses Enter, which is similar to how other Linux CLI tools work

Also noteworthy is that the `-y` option avoids the need for user confirmation in the situations above, causing a remote to be chosen automatically and a similar ref to be chosen automatically, but only if there's only one match (if more than one ref is similar, the operation is aborted). It's expected that scripts will always specify the remote name and full app ID, so this should only affect users who choose to use `-y` on the command line.

cc: @allanday

### CODE REVIEW

**Comment by mwleeds** [comment]
This seems to work but is not ready to be merged yet. I still need to:

- [ ] make it work with the `--no-pull` option
- [ ] make it exclude things like `.Debug` and `.Sources` extensions by default (unless they're the only match)
- [ ] see if it's possible to make the logic in `find_matching_refs()` cleaner

**Comment by mwleeds** [comment]
This seems to be working and I think it's ready for review. I will also test it more since I've only so far tested it cursorily.

**Comment by mwleeds** [comment]
bot, retest this please

**Comment by mwleeds** [comment]
bot, retest this please

**Comment by matthiasclasen** [comment]
I tried flathub install devhelp, and it asked me:

$ ./flatpak install devhelp
Found ref ‘devhelp’ in remote ‘flathub’ (user). Is that the remote you want to use? [Y/n]:

which I find less than ideal, in two ways: 
- it doesn't spell out the actual ref that it is going to install
- it forces me to make a decision about what remote to use before I see what is actually available

File: common/flatpak-utils-private.h [hunk 1]
```diff
@@ -676,6 +676,8 @@ gboolean flatpak_check_required_version (const char *ref,
                                          GKeyFile   *metakey,
                                          GError    **error);
 
+int flatpak_levenshtein_distance (const char *s, const char *t);
+
```
Comment (matthiasclasen) [inline common/flatpak-utils-private.h]: It would be nice to do this move as a preparatory commit, separately

File: doc/flatpak-install.xml [hunk 2]
```diff
@@ -66,8 +66,11 @@
             flatpak ref format, which looks like "(app|runtime)/ID/ARCH/BRANCH". All elements
             except ID are optional and can be left out, including the slashes,
             so most of the time you need only specify ID. Any part left out will be matched
-            against what is in the remote, and if there are multiple matches an error message
-            will list the alternatives.
+            against what is in the remote, and if there are multiple matches you will be
+            prompted to choose from them. You will also be prompted with choices if
```
Comment (matthiasclasen) [inline doc/flatpak-install.xml]: Would 'choose among them' or 'choose one of them' be better ?

File: doc/flatpak-install.xml [hunk 3]
```diff
@@ -33,7 +33,7 @@
             <cmdsynopsis>
                 <command>flatpak install</command>
                 <arg choice="opt" rep="repeat">OPTION</arg>
-                <arg choice="plain">REMOTE-NAME</arg>
+                <arg choice="opt">REMOTE-NAME</arg>
```
Comment (matthiasclasen) [inline doc/flatpak-install.xml]: there is a (preexisting) mismatch here between REMOTE-NAME in the display and REMOTE in the paragraph. We should fix that at the same time, maybe ?

**Comment by mwleeds** [comment]
> I tried flatpak install devhelp, and it asked me:
> 
> $ ./flatpak install devhelp
> Found ref ‘devhelp’ in remote ‘flathub’ (user). Is that the remote you want to use? [Y/n]:
> 
> which I find less than ideal, in a few ways:
> 
> * it doesn't spell out the actual ref that it is going to install. "Found ref 'devhelp'" is arguable a lie - devhelp is not a ref.

Sure we could change it to the full ref
> * it forces me to make a decision about what remote to use before I see what is actually available

As I said in the commit message, one reason we might not want to search every configured remote for the given partial ref  and let the user choose between them is the performance cost of that, but I haven't measured that.
> I would have expected output more similar to what we say if a remote is specified:
> 
> Similar refs found for ‘devhelp’:
> 
> 1. app/org.gnome.Devhelp/x86_64/stable in remote 'flathub'
> 2. app/org.gnome.Devhelp.nightly/x86_64/master in remote 'gnome-apps-nightly'
>    ...

I can see the appeal of having one prompt instead of two but there is at least one reason we might not want to do it that way:

 We only support installing from one remote into one installation, so if the user specified multiple partial refs on the command line, after they choose a `(full ref, remote)` pair for the first one we'd have to restrict where we search for the rest of them, which could be counterintuitive behavior. It would be easy to add the ability to install from different remotes in the same installation, but I'm not sure we want to have `install` create more than one transaction for more than one installation.

**Comment by matthiasclasen** [comment]
> As I said in the commit message, one reason we might not want to search every configured remote for the given partial ref and let the user choose between them is the performance cost of that, but I haven't measured that.

If this takes a long time, I think a better approach would be to print candidates as we find them, and let the user interrupt using Escape, and choose between the matches found so far (that may not be easy to implement, of course).

> We only support installing from one remote into one installation, so if the user specified multiple partial refs on the command line, after they choose a `(full ref, remote)` pair for the first one we'd have to restrict where we search for the rest of them, which could be counterintuitive behavior. It would be easy to add the ability to install from different remotes in the same installation, but I'm not sure we want to have `install` create more than one transaction for more than one installation.

I think using one transaction per installation is fine, if that is what we have to do. But really, installing multiple refs by search seems like a corner case to me, and if it was down to this, I would suggest to only allow install/search for a single ref.

**Comment by matthiasclasen** [comment]
I think a minimal change I'd like to see is this:

https://github.com/matthiasclasen/flatpak/commit/3dff45eb63c9a07ea6c678dd6c5f01db8ba2926e

Keep going if the first remote isn't selected. Is there any problem with this?

**Comment by mwleeds** [comment]
> I think a minimal change I'd like to see is this:
> 
> [matthiasclasen@3dff45e](https://github.com/matthiasclasen/flatpak/commit/3dff45eb63c9a07ea6c678dd6c5f01db8ba2926e)
> 
> Keep going if the first remote isn't selected. Is there any problem with this?

Sure, I squashed that change in

**Comment by matthiasclasen** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**